src/trues.jl

```diff
@@ -0,0 +1,38 @@
+
+"""
+    Trues = Ones{Bool, N, Axes} where {N, Axes}
+
+Lazy version of `trues` with axes.
+Typically created using `Trues(dims)` or `Trues(dims...)`
+
+# Example
+```jldoctest
+julia> Trues(1,3)
+1×3 Ones{Bool,2,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}}} = true
+
+julia> Trues((2,3))
+2×3 Ones{Bool,2,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}}} = true
+```
+"""
+const Trues = Ones{Bool, N, Axes} where {N, Axes}
+
+
+""" `Falses = Zeros{Bool, N, Axes}` (see `Trues`) """
+const Falses = Zeros{Bool, N, Axes} where {N, Axes}
+
+
+# y[mask] = x when mask isa Trues (cf y[:] = x)
+function Base.setindex!(y::AbstractArray{T,N}, x, mask::Trues{N}) where {T,N}
+    @boundscheck axes(x) == axes(mask) == axes(y) ||
+        throw(DimensionMismatch("axes must match for setindex!"))
+    @boundscheck checkbounds(y, mask)
+    copyto!(y, x)
+end
+
+# x[mask] when mask isa Trues (cf x[trues(size(x))] or x[:])
+function Base.getindex(x::AbstractArray{T,D}, mask::Trues{D}) where {T,D}
+    @boundscheck axes(x) == axes(mask) ||
+        throw(DimensionMismatch("axes must match for getindex"))
```

Comment:
This should be a `BoundsError` to match `Array`:
```julia
julia> a = randn(5)
5-element Array{Float64,1}:
  0.5894229873306978
  1.1656206737590917
  0.314932122096189
 -0.8914119854410714
  0.7630084129166993

julia> a[[true,true]]
ERROR: BoundsError: attempt to access 5-element Array{Float64,1} at index [Bool[1, 1]]
Stacktrace:
 [1] throw_boundserror(::Array{Float64,1}, ::Tuple{Base.LogicalIndex{Int64,Array{Bool,1}}}) at ./abstractarray.jl:541
 [2] checkbounds at ./abstractarray.jl:506 [inlined]
 [3] _getindex at ./multidimensional.jl:742 [inlined]
 [4] getindex(::Array{Float64,1}, ::Array{Bool,1}) at ./abstractarray.jl:1060
 [5] top-level scope at REPL[2]:1
```

Reply:
done